libafl_bolts/src/lib.rs

```diff
@@ -1050,44 +1050,19 @@ pub unsafe fn set_error_print_panic_hook(new_stderr: RawFd) {
     }));
 }
 
-// Credit goes to https://github.com/thomcc/nonzero_lit
-// We don't want add another dependency and just want to use usize macro of it.
-#[doc(hidden)]
-pub mod _private {
-    pub use core::num::NonZeroUsize;
-
-    macro_rules! define_nz_ctor {
-        ($(pub fn $nz_func:ident($n:ident : $int:ident) -> $NonZeroInt:ident;)+) => {$(
-            #[inline]
-            #[must_use]
-            pub const fn $nz_func($n : $int) -> $NonZeroInt {
-                // Note: Hacky const fn assert.
-                let _ = ["N must not be zero"][($n == 0) as usize];
-
-                match $NonZeroInt::new($n) {
-                    Some(x) => x,
-                    // The assert above makes this branch unreachable
-                    None => unreachable!(),
-                }
-            }
-        )+};
-    }
-
-    define_nz_ctor! {
-        pub fn nz_usize(n: usize) -> NonZeroUsize;
-    }
-}
-
-/// 0 cost way to create check nonzero on compilation.
+/// Zero-cost way to construct [`core::num::NonZeroUsize`] at compile-time.
 #[macro_export]
 macro_rules! nonzero {
-    ($val:expr $(,)?) => {{
-        const __E: usize = $val;
-        {
-            const NZ: $crate::_private::NonZeroUsize = $crate::_private::nz_usize(__E);
-            NZ
+    // TODO: Further simplify with `unwrap`/`expect` once MSRV includes
+    // https://github.com/rust-lang/rust/issues/67441
+    ($val:expr) => {
+        const {
+            match core::num::NonZeroUsize::new($val) {
+                Some(x) => x,
+                None => panic!("Value passed to `nonzero!` was zero"),
```

Comment:
Does this panic at compile time though? Or just insert a panici at runtime?

Reply:
> Does this panic at compile time though? Or just insert a panici at runtime?

It's my impression that everything in a `const` block happens at compile-time. See [the docs](https://doc.rust-lang.org/reference/const_eval.html#constant-expressions)

> Certain forms of expressions, called constant expressions, can be evaluated at compile time. In [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context), these are the only allowed expressions, and are always evaluated at compile time.

> A const context is one of the following:
> [...]
> A [const block](https://doc.rust-lang.org/reference/expressions/block-expr.html#const-blocks)

> FWIW If it panics at compile time we should just make this a fn not a macro

This could certainly be a `const fn`, I think the key would be that all call-sites would need to surround the call in a `const` block (as per the above). Otherwise, you could accidentally pass in non-`const` arguments and get a runtime panic.

Reply:
See my last example